parser.py

```diff
@@ -64,129 +154,85 @@ def parseMeta(line, recipe):
-                if start == '!' or start == '+' or not line:
-                    state = State.FinishPhase
-                    continue
-
-                if start == '*':
-                    state = State.Steps
-                    continue
-                
-                if start == '#':
-                    phase.ingredients.append(parseIngredient(line))
-
-            elif state == State.Steps:
-                if start == '!' or start == '#' or start == '+' or not line:
-                    state = State.FinishPhase
-                    continue
-
-                if start == '*':
-                    line = line.lstrip('* \t')
-                    line = line.rstrip(' \t\n')
-                    phase.steps.append(Step(line))
-
-            elif state == State.WaitPhase:
-                if start == '!' or start == '#' or start == '*' or not line:
-                    recipe.phases.append(waitphase)
-                    waitphase = WaitPhase()
-                    if start == '!' or not line:
-                        state = State.FinishRecipe
-                    elif start == '#':
-                        state = State.Ingredients
-                    elif start == '*':
-                        state = State.Steps
-                    continue
-                elif start == '+':
-                    parseWaitPhase(line, waitphase)
-           
-            elif state == State.FinishPhase:
-                recipe.phases.append(phase)
-                phase = Phase()
-                if start == '!' or not line:
-                    state = State.FinishRecipe
-                elif start == '#':
-                    state = State.Ingredients
-                elif start == '+':
-                    state = State.WaitPhase
-                continue
-
-            elif state == State.FinishRecipe:
-                recipes.append(recipe)
-                recipe = Recipe()
-                state = State.Waiting
-                continue
-
-            line = stream.readline()
-            if not line and state == State.Waiting:
-                break
+        recipe = None
+        phase = None
+        meta = False
+        for line in preprocessLines(stream):
+            if line.lineType == Line.Command:
+                key, value = splitCommand(line.contents)
+                if recipe and key == 'title':
+                    recipe = None
+                if not recipe:
+                    if key != 'title':
+                        raise Exception('recipes must start with a title')
+                    recipe = Recipe()
+                    recipes.append(recipe)
+                    meta = True
+                if meta:
+                    parseMeta(key, value, recipe)
+                else:
+                    raise Exception('commands are not yet used (metadata must appear at the beginning of the recipe)')
+
+            elif line.lineType == Line.Plain:
+                if len(recipe.phases):
```

Comment:
True, so it should be "if recipe is None or len(recipe.phases):".
I can't use meta since empty lines will set meta to False, so I can separate and !img at the end of the metadata and an !img at the start of the contents as a command to add a progress picture (which I still plan to implement).